### ISSUE TITLE
Post a new transaction on the network.

### ISSUE THREAD

**Issue by Highjhacker** [Highjhacker/Ark-Elixir#7]
So, I think that we need to mention the elephant in the room ; creating a new transaction.

It's something needed, so if you have any thoughts on the subject, let's discuss it !

I guess the first thing to do is to check how it's implemented in [Ark-js](https://github.com/ArkEcosystem/ark-js/) and check what we can do with elixir.

[Swagger Doc](https://ark.brianfaust.me/#/Transport/transport_addTransactions)

**Comment by sudomilk** [comment]
From looking at swagger, I am going to assume that this is what a proper transaction object looks like in JSON:

```json
{
      "id": "e780c04ddd37cfb603528255d3d1c6a5723d4a4fea599a0576ee54577e7bb6d6",
      "blockid": "1000000289854238649",
      "type": 0,
      "timestamp": 22231900,
      "amount": 999990000000,
      "fee": 10000000,
      "senderId": "AWHdJfbh7KchdyFagJXodKuobJSVLgM4P1",
      "recipientId": "AUexKjGtgsSpVzPLs6jNMM6vJ6znEVTQWK",
      "senderPublicKey": "032a5554f5dd6604b9be8e85cfa7cf6dbbc9fe943cd486a270504ff98c98ee4efb",
      "signature": "3045022100e665ac784320da891df1c62a890fd7ef2a2a0992d2fca2a9c5293974e5304caa02202682796b088961e4377c7004b2a74a9e9db53034b027f936349dd428de8fd210",
      "asset": {},
      "confirmations": 294743
    }
```

My suggestion would be to create a new module, called `Ark_Elixir.Transaction`. In there, we can create a struct to match this object. We can then add some functions if we want like to validate or create a new transaction object. This would also give us a new type called `Transaction.t` that we could use in typespecs!

From there, it should be fairly straightforward to just send a post request with this transaction object.

**Comment by sudomilk** [comment]
Ah, okay so I'm prototyping something and see what you mean. For the transaction object, they don't have a lot of info about what is valid. Looking into it.

**Comment by insanedefaults** [comment]
Great issue! The transaction protocol is being updated. The details are outlined in [AIP 11](https://github.com/ArkEcosystem/AIPs/blob/master/AIPS/aip-11.md).

You may have noticed that the Swagger UI has a deprecated post transaction and a new one under peer. The ark-js project has an implementation of the new protocol [here](https://github.com/ArkEcosystem/ark-js/commit/95b591c4641ab12869890889c9e3cbbcbe457166).

I suppose the question we should first answer is this do we **support the old and new protocols** or **support only the new protocol**?

From what I am seeing, the old protocol is still in use and the AIP for the new protocol still is in open discussion. I think implementing the old one now would be best. That said, I have no idea how to sign the transaction or anything, so it will likely require research to build a decent prototype. If you know more, that would be great!

**Comment by Highjhacker** [comment]
Agree, we should use the old/current one for now and then implement the new one when needed.

For making a new transaction, what I'm guessing is, you have to generate the correct json response, which can be easy for things like the timestamp of the transaction etc. But then some things are more complexes, like, the new transaction should have a valid TXID, but how to generate it to be in compliance with Ark ? 

So I guess we should start looking at which crypto libs in Elixir can help us to achieve this and how the current process is implemented in the others Ark wrapper. (Arky, Ark-JS and SwiftyArk, ..)

**Comment by insanedefaults** [comment]
Sure thing, if you want, we can use this issue to post what we find.

**Comment by Highjhacker** [comment]
Gonna post the informations I'm gathering on how to create a valid transaction and edit this when I find something new.

[This](https://github.com/ArkEcosystem/ark-js/blob/0b8d460a54657b24703e0b7fe59cf2067e00491b/lib/transactions/transaction.js#L16) is what we need I guess.

- We need to check if the address provided in the transaction is valid (See [here](https://github.com/ArkEcosystem/ark-js/blob/0b8d460a54657b24703e0b7fe59cf2067e00491b/lib/v2/transactions/crypto.js#L314) )

- We need to check the public key

- We need to generate a valid txID

- Optional vendorfield

- Validate the password (and the second one too if it exists)

- Calculate the correct timestamp (see [this](https://gist.github.com/Highjhacker/83fa18cb80f48a7b31bd07f655830fba) and tell me what you think about that)

- ...

- Then we need to wrap it up and POST it to the correct [endpoint](https://ark.brianfaust.me/#/Transport/transport_addTransactions)

**Comment by insanedefaults** [comment]
If the timex diff is just returning an epoch difference maybe we could skip around just using the library and calculate `current_epoch_time - 1490187600`.
  
I'm going to dig into putting [the validateaddress](https://github.com/ArkEcosystem/ark-js/blob/0b8d460a54657b24703e0b7fe59cf2067e00491b/lib/transactions/crypto.js#L494) function in Elixir later tonight.

**Comment by Highjhacker** [comment]
Hm, clever, won't change much and should be better performances wise, less dependencies more freedom 😄 Gonna look at that.

**Comment by Highjhacker** [comment]
[Without Timex](https://gist.github.com/Highjhacker/8676ffcc017458d8bacf739a0cfea6b6)
